posthog/temporal/batch_exports/http_batch_export.py

```diff
@@ -0,0 +1,361 @@
+import datetime as dt
+import json
+from dataclasses import dataclass
+
+from django.conf import settings
+from temporalio import activity, workflow
+from temporalio.common import RetryPolicy
+
+from posthog.batch_exports.service import BatchExportField, BatchExportSchema, HttpBatchExportInputs
+from posthog.temporal.batch_exports.base import PostHogWorkflow
+from posthog.temporal.batch_exports.batch_exports import (
+    BatchExportTemporaryFile,
+    CreateBatchExportRunInputs,
+    UpdateBatchExportRunStatusInputs,
+    create_export_run,
+    execute_batch_export_insert_activity,
+    get_data_interval,
+    get_rows_count,
+    iter_records,
+    json_dumps_bytes,
+)
+from posthog.temporal.batch_exports.clickhouse import get_client
+from posthog.temporal.batch_exports.metrics import (
+    get_bytes_exported_metric,
+    get_rows_exported_metric,
+)
+from posthog.temporal.common.logger import bind_temporal_worker_logger
+
+
+class RetryableResponseError(Exception):
+    """Error for HTTP status >=500 (plus 429)."""
+
+    def __init__(self, status):
+        super().__init__(f"RetryableResponseError status: {status}")
+
+
+class NonRetryableResponseError(Exception):
+    """Error for HTTP status >= 400 and < 500 (excluding 429)."""
+
+    def __init__(self, status):
+        super().__init__(f"NonRetryableResponseError status: {status}")
+
+
+def raise_for_status(response: aiohttp.ClientResponse):
+    """Like aiohttp raise_for_status, but it distinguishes between retryable and non-retryable
+    errors."""
+    if not response.ok:
+        if response.status >= 500 or response.status == 429:
+            raise RetryableResponseError(response.status)
+        else:
+            raise NonRetryableResponseError(response.status)
+
+
+def http_default_fields() -> list[BatchExportField]:
+    return [
+        BatchExportField(expression="toString(uuid)", alias="uuid"),
+        BatchExportField(expression="timestamp", alias="timestamp"),
+        BatchExportField(expression="COALESCE(inserted_at, _timestamp)", alias="_inserted_at"),
+        BatchExportField(expression="event", alias="event"),
+        BatchExportField(expression="nullIf(properties, '')", alias="properties"),
+        BatchExportField(expression="toString(distinct_id)", alias="distinct_id"),
+        BatchExportField(expression="elements_chain", alias="elements_chain"),
+    ]
+
+
+class HeartbeatDetails:
+    """This class allows us to enforce a schema on the Heartbeat details.
+
+    Attributes:
+        last_uploaded_part_timestamp: The timestamp of the last part we managed to upload.
+    """
+
+    last_uploaded_part_timestamp: str
+
+    def __init__(self, last_uploaded_part_timestamp: str):
+        self.last_uploaded_part_timestamp = last_uploaded_part_timestamp
+
+    @classmethod
+    def from_activity_details(cls, details) -> "HeartbeatDetails":
+        last_uploaded_part_timestamp = details[0]
+        return HeartbeatDetails(last_uploaded_part_timestamp)
```

Comment:
Bit of copy pasting: We are not uploading "parts" here, that's a term from S3 multipart uploads.